capstone-backend/src/main/java/com/google/sps/servlets/SearchServlet.java

```diff
@@ -0,0 +1,181 @@
+import com.google.gson.JsonArray;
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.sps.data.VolumeData;
+import java.io.BufferedReader;
+import java.io.InputStreamReader;
+import java.io.IOException;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.util.Map;
+import java.util.ArrayList;
+import java.util.Collection;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/** Servlet that returns search data from the Google Books API. */
+@WebServlet("/api/search")
+public class SearchServlet extends HttpServlet {
+
+  public final static int DEFAULT_MAX_RESULTS = 5;
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String fullOutput = "";
+    String searchTerm = request.getParameter("searchTerm");
+    int maxResults = parseNaturalNumber(request.getParameter("maxResults"));
+    if (maxResults == -1) {
+      maxResults = DEFAULT_MAX_RESULTS;
+    }
+    
+    if (searchTerm == null || searchTerm.isEmpty()) return;
+    searchTerm = URLEncoder.encode(searchTerm, "UTF-8");
+
+    try {
+      String formattedURL = String.format("https://www.googleapis.com/books/v1/volumes?q={%s}&maxResults=%d&country=US",
+        searchTerm, maxResults);
+		  URL url = new URL(formattedURL);
+		  HttpURLConnection conn = (HttpURLConnection) url.openConnection();
+		  conn.setRequestMethod("GET");
+		  conn.setRequestProperty("Accept", "application/json");
+
+      if (conn.getResponseCode() != 200) {
+        throw new RuntimeException("Failed : error code : "
+            + conn.getResponseCode());
+      }
+      BufferedReader br = new BufferedReader(new InputStreamReader(
+        (conn.getInputStream())));
+      
+      String output;
+      while ((output = br.readLine()) != null) {
+        fullOutput += output;
+      }
+      conn.disconnect();
+    } catch (MalformedURLException e) {
+      e.printStackTrace();
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+    
+    response.setContentType("application/json;");
+    Gson gson = new GsonBuilder().disableHtmlEscaping().create();
+
+    response.getWriter().println(gson.toJson(convertResponseToVolumeData(fullOutput)));
+  }
+
+  /*
+   * Takes in a response from the Google Books API and converts it to a Collection of VolumeData objects.
+   * @param output the response from the Books API
+   * @return a Collection of the books from the API call, in VolumeData form
+   */
+  @VisibleForTesting
+  static Collection<VolumeData> convertResponseToVolumeData(String output) {
+    ArrayList<VolumeData> volumes = new ArrayList<>();
+
+    // Parse data into json object
+    JsonParser parser = new JsonParser();
```

Comment:
According to Gitpod, ```JsonParser``` is deprecated.

Reply:
I see, looks like I can just call JsonParser statically now and not create a new object. I will look into it!